### DISCUSSION TITLE
Looking for additional insight into Keil C[X]51 toolchain object and library file formats

### DISCUSSION THREAD

**Discussion by gtackett** [NationalSecurityAgency/ghidra#4046] [Q&A]
ARM Keil's toolchain for devices in the Intel 8051 family lineage is essentially Intel's OMF51 format with Keil's own extensions. For these extensions, Keil only offers documentation to developers for the processor family on a case by case basis. At least Intel has made its own basic OMF51 documentation available to the public at no cost.

My employer's client is **not** developing for that device, so it's very hard to predict how Keil might respond to a request for this documentation. I lack any official capacity or authority to approach them or anyone specific about getting this documentation. Yet it would be very helpful to have for some of the reverse engineering we are doing.

Our client, my own employer, and I personally, are all honor bound not to violate any protections of IP or other information that Keil and its country of incorporation chooses to impose. I haven't yet asked Keil what those may be, but I will probably pose to them soon the hypothetical (that is, completely unofficial) situation to see if they'll spell out their policy re. protection of this documentation.

I also wish to be clear that I want anyone I communicate with to honor both Keil's restrictions (if they are known to them) and their own employer's (if applicable) code of ethics and any governing legal code where they reside.

**With that disclaimer understood:**

What I need is an analyzer for Keil's object and library files, including their extensions to the Intel original format,  that will be useful in reverse engineering binaries for that platform.

I've already written a [**010 Editor**](https://www.sweetscape.com) template that implements the bare bones of OMF51 file format.
If someone who's following Ghidra has the OMF51 extensions information already available and is not legally or ethically bound to protect it, they might save us considerable time and expense by making it available to the interested public.

Such knowledge would also enable the development of Ghidra binary loaders, user scripts, etc., that might be of benefit to others interested in this device family.

--
These comments are my own entirely, and not necessarily those of my employer or our client.

**Comment by gtackett** [comment]
My OMF51 template for 010 Editor is now available in [Sweetscape's template library](https://www.sweetscape.com/010editor/repository/templates/).  It's called OMF51.bt. I hope someone here can have a look and perhaps make or suggest improvements.

(It only deals with object files produced by Intel’s original toolchain for its original 8051. It also doesn’t attempt to deal with the processor’s instructions—it just recognizes the object file format in which they’re wrapped.)

**Comment by ABratovic** [comment]
Hi,
Navigate to https://www.keil.com/download/list/C251.htm
click on 'MCB251SB RAM Test' and download RAMTEST.zip and unzip the files compiled 30 years ago!!!

I found the 80251 RAMTEST.obj file crashes the Ghidra OMF-51 loader ☹️
I also noted the RT.HEX file when loaded into Ghidra has a different test string to what is shown in all the other files. :smiley:

Could you please test ramtest.obj against your 010 template because when I used a VM not connected to the internet I got
Executing template 'C:\xxx\OMF51.bt' on 'Z:\kiel_251\ramtest\RAMTEST.OBJ'...
*ERROR Line 481: Unsupported value for custom attribute 'hidden'. 

I also got the same error trying to view obj files I created 15 years ago by the Raisonance Ride7 8051 compiler for a SiLabs derivative, 
so am unsure if it's how I installed the template on my VM or the compiler suite?

**Reply by ryanmkurtz** [reply]
> I found the 80251 RAMTEST.obj file crashes the Ghidra OMF-51 loader 

i'm working on a fix for this now

**Reply by ryanmkurtz** [reply]
I have it loading and it identfies the `DEPLST 0x70` record properly, but we don't recognize any of the other record types:

For example, we don't recognize record numbers `0xc0`, `0xc8`, `0x9a`, etc.

<img width="796" height="784" alt="image" src="https://github.com/user-attachments/assets/49c7a148-67be-4385-b26f-dbf014043fa2" />

This is what we know about:
https://github.com/NationalSecurityAgency/ghidra/blob/cccc5103c13d8358b7dd6564ad325f80a9060336/Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/omf/omf51/Omf51RecordTypes.java#L27-L60
https://github.com/NationalSecurityAgency/ghidra/blob/cccc5103c13d8358b7dd6564ad325f80a9060336/Ghidra/Features/Base/src/main/java/ghidra/app/util/bin/format/omf/omf166/Omf166RecordTypes.java#L25-L62

**Reply by ryanmkurtz** [reply]
The binary will now load in the `master` branch, and `Ghidra 12.1` (when it comes out).

**Reply by gtackett** [reply]
> Could you please test ramtest.obj against your 010 template

My 010 template only handles the original Intel OMF-51 object file format and Keil’s extensions to that for their own toolchain. It knows nothing about any instruction set, but this is moot anyway since your issue has been resolved.

**Comment by ABratovic** [comment]
Awesome, hopefully @gtackett can chime in with more details.
Also, when decompiling 80251 RT.HEX, to resolve RAM:000347 a5 ??

      RAM:000344 a5  78  f6       JNE        LAB_RAM_00033d
      RAM:000347 a5              ??         A5h
      RAM:000348 7e              ??         7Eh    ~
      RAM:000349 f8              ??         F8h
      RAM:00034a 00              ??         00h
      RAM:00034b 3d              ??         3Dh    =
      RAM:00034c 02  03  8a       LJMP       LAB_RAM_00038a


in 80251.sinc change
```
# NOTE: must use constraints DRK, DRKD and DRKS
attach variables  [ drk47 drk03 drk47_ drk03_ ] [
	DR0  DR4  DR8  DR12 DR16 DR20 DR24 DR28
	DPX  SPX  _    _    _    _    _    _
];
```
to
```
# NOTE: must use constraints DRK, DRKD and DRKS
# DR56 = DPX, DR60 = SPX
attach variables  [ drk47 drk03 drk47_ drk03_ ] [
    DR0  DR4  DR8  DR12 DR16 DR20 DR24 DR28
    _    _    _    _    _    _    DPX  SPX
];
```
to get 
```A5 7E F8 00 3D                    mov     SPX, #0x3D```

**Reply by GhidorahRex** [reply]
That last change is resolved with the fix for #8395 in 12.0.3

**Reply by ABratovic** [reply]
> That last change is resolved with the fix for #8395 in 12.0.3

I missed that one, I'm using v12.0 Public
